File: Server/Tuckshop.Domain/Tuckshop.Core.Models/Models/Product.cs
namespace Tuckshop.Core.Models
{
  using Microsoft.EntityFrameworkCore;
  using Neo.Model;
  using Neo.Model.Validation;
  using System.ComponentModel.DataAnnotations;
  using System.ComponentModel.DataAnnotations.Schema;

  /// <summary>
  /// A class representing a Product
  /// </summary>
  public class Product : ModelBase<Product>
  {
    /// <summary>
    /// Gets or sets the Product Id
    /// </summary>
    public int ProductId { get; set; }

    /// <summary>
    /// Gets or sets the Product Name
    /// </summary>
    [Required]
    [MaxLength(100)]
    public string ProductName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the Price  
    /// </summary>
    [Column(TypeName = "money")]
    public decimal Price { get; set; }

    /// <inheritdoc />
    protected override void AddBusinessRules(ValidationRules<Product> rules)
    {
      base.AddBusinessRules(rules);

      rules.FailWhen(c => c.Price <= 0, "Price must be above zero.");
    }
  }
}

File: Client/Tuckshop.WebApp/src/Domain/Models/Product.ts
import { Attributes, ModelBase, Rules, Validation } from '@singularsystems/neo-core';

export default class Product extends ModelBase {
    static typeName = "Product";

    constructor() {
        super();
        this.makeObservable();
    }

    public productId: number = 0;

    @Rules.Required()
    @Rules.StringLength(100)
    public productName: string = "";

    @Attributes.Float()
    @Rules.Required()
    public price: number = 0;

    @Rules.StringLength(300)
    public imageUrl: string = "";

    // Client only properties / methods

    protected static addBusinessRules(rules: Validation.Rules<Product>) {
        super.addBusinessRules(rules);
        rules.failWhen(c => c.price <= 0, "Price must be above zero.");
    }

    public toString(): string {
        if (this.isNew || !this.productName) {
            return "New product";
        } else {
            return this.productName;
        }
    }
}